### PR TITLE
Add support for invalidation by subgraph name and/or type name

### PR DESCRIPTION
**Pull request by garypen**
Scan redis using supplied subgraph and entity type name and delete the matching keys.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/plugins/cache/entity.rs [hunk 1]
```diff
@@ -539,6 +539,7 @@ async fn cache_lookup_root(
 
     let key = extract_cache_key_root(
         &name,
+        None,
```
Comment (garypen) [inline apollo-router/src/plugins/cache/entity.rs:542]: Question for reviewers: Is there a way to figure out the entity type here?

File: apollo-router/src/plugins/cache/entity.rs [hunk 2]
```diff
@@ -842,14 +845,17 @@ fn extract_cache_key_root(
...
-    // - subgraph name: caching is done per subgraph
+    // - subgraph name: subgraph name
+    // - entity type: entity type
     // - query hash: invalidate the entry for a specific query and operation name
     // - additional data: separate cache entries depending on info like authorization status
     let mut key = String::new();
     let _ = write!(
         &mut key,
-        "subgraph:{subgraph_name}:Query:{query_hash}:{additional_data_hash}"
+        "subgraph:{subgraph_name}:type:{entity_type}:hash:{query_hash}:data:{additional_data_hash}"
```
Comment (garypen) [inline apollo-router/src/plugins/cache/entity.rs:858]: drive-by: fixing to match changes in key structure in spec.

File: apollo-router/tests/integration/redis.rs [hunk 3]
```diff
@@ -416,7 +416,7 @@ async fn entity_cache() -> Result<(), BoxError> {
     insta::assert_json_snapshot!(response);
 
     let s:String = client
-          .get("subgraph:products:Query:0df945dc1bc08f7fc02e8905b4c72aa9112f29bb7a214e4a38d199f0aa635b48:d9d84a3c7ffc27b0190a671212f3740e5b8478e84e23825830e97822e25cf05c")
+          .get("subgraph:products:type:Query:hash:0df945dc1bc08f7fc02e8905b4c72aa9112f29bb7a214e4a38d199f0aa635b48:data:d9d84a3c7ffc27b0190a671212f3740e5b8478e84e23825830e97822e25cf05c")
```
Comment (garypen) [inline apollo-router/tests/integration/redis.rs:419]: drive-by: fixing to match changes in key structure in spec.

File: apollo-router/tests/integration/redis.rs [hunk 4]
```diff
@@ -743,7 +743,7 @@ async fn entity_cache_authorization() -> Result<(), BoxError> {
...
```
Comment (garypen) [inline apollo-router/tests/integration/redis.rs:746]: drive-by: fixing to match changes in key structure in spec.

File: apollo-router/src/plugins/cache/entity.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/plugins/cache/entity.rs:542]: * get the schema at plugin creation (there's a different plugin init function signature for that)
* get the API schema which is a `apollo_copiler::Schema`
* `schema.schema_definition.query.as_ref().map(|q| q.name.to_string())`

**Review by Geal — CHANGES_REQUESTED** [review]
could you split that test from the subgraph level invalidation test?

**Comment by garypen** [comment]
> could you split that test from the subgraph level invalidation test?

done: 5256aff5ea5cf1d8ad47b78202bb25277e0554d1

**Review by Geal — APPROVED** [review]
